### PR TITLE
Add "new window" option to news forwardings

### PR DESCRIPTION
**Pull request by LazuLitee**

### CODE REVIEW

File: core_modules/News/Controller/News.class.php [hunk 1]
```diff
@@ -351,7 +353,9 @@ private function getDetails()
             if ($this->_objTpl->blockExists('news_text')) {
                 $this->_objTpl->parse('news_text');
             }
-            if ($this->_objTpl->blockExists('news_redirect')) {
+            if ($this->_objTpl->blockExists('news_redirect') && $newWindow) {
```
Comment (daeppen) [inline core_modules/News/Controller/News.class.php]: what's the purpose of this change?

File: core_modules/News/Controller/News.class.php [hunk 2]
```diff
@@ -705,8 +709,9 @@ private function parseRelatedMessagesOfMessage($messageId, $relatedByKind, $rela
                                     : \Cx\Core\Routing\Url::fromModuleAndCmd('News', $this->findCmdById('details', array_keys($newsCategories)), FRONTEND_LANG_ID, array('newsid' => $newsid)))
                                 : $objResult->fields['redirect'];
 
-            $htmlLink       = self::parseLink($newsUrl, $newstitle, contrexx_raw2xhtml('['.$_ARRAYLANG['TXT_NEWS_MORE'].'...]'));
-            $htmlLinkTitle  = self::parseLink($newsUrl, $newstitle, contrexx_raw2xhtml($newstitle));
+            $newWindow      = empty($objResult->fields['newWindow']) ? "_self" : "_blank";
```
Comment (daeppen) [inline core_modules/News/Controller/News.class.php]: Use single quotes (instead of double quotes) for simple, non-functional strings

File: core_modules/News/Controller/News.class.php [hunk 3]
```diff
@@ -1037,6 +1044,7 @@ private function getHeadlines() {
                    'NEWS_LINK_TITLE'     => $htmlLinkTitle,
                    'NEWS_LINK'           => $htmlLink,
                    'NEWS_LINK_URL'       => contrexx_raw2xhtml($newsUrl),
+                   'NEWS_LINK_TARGET'    => empty($objResult->fields['newWindow']) ? "_self" : "_blank",
```
Comment (daeppen) [inline core_modules/News/Controller/News.class.php]: Use single quotes (instead of double quotes) for simple, non-functional strings

File: core_modules/News/Controller/News.class.php [hunk 4]
```diff
@@ -1258,8 +1266,9 @@ private function getTopNews() {
                                         : \Cx\Core\Routing\Url::fromModuleAndCmd('News', $this->findCmdById('details', array_keys($newsCategories)), FRONTEND_LANG_ID, array('newsid' => $newsid)))
                                     : $objResult->fields['redirect'];
 
-                $htmlLink       = self::parseLink($newsUrl, $newstitle, contrexx_raw2xhtml('['.$_ARRAYLANG['TXT_NEWS_MORE'].'...]'));
-                $htmlLinkTitle  = self::parseLink($newsUrl, $newstitle, contrexx_raw2xhtml($newstitle));
+                $newWindow      = empty($objResult->fields['newWindow']) ? "_self" : "_blank";
```
Comment (daeppen) [inline core_modules/News/Controller/News.class.php]: Use single quotes (instead of double quotes) for simple, non-functional strings

File: core_modules/News/Controller/News.class.php [hunk 5]
```diff
@@ -1283,6 +1292,7 @@ private function getTopNews() {
                    'NEWS_LINK_TITLE'    => $htmlLinkTitle,
                    'NEWS_LINK'          => $htmlLink,
                    'NEWS_LINK_URL'      => contrexx_raw2xhtml($newsUrl),
+                   'NEWS_LINK_TARGET'   => empty($objResult->fields['newWindow']) ? "_self" : "_blank",
```
Comment (daeppen) [inline core_modules/News/Controller/News.class.php]: Use single quotes (instead of double quotes) for simple, non-functional strings

File: core_modules/News/Controller/News.class.php [hunk 6]
```diff
@@ -1005,8 +1011,9 @@ private function getHeadlines() {
                                     )
                                     : $objResult->fields['redirect'];
 
-                $htmlLink       = self::parseLink($newsUrl, $newstitle, contrexx_raw2xhtml('['.$_ARRAYLANG['TXT_NEWS_MORE'].'...]'));
-                $htmlLinkTitle  = self::parseLink($newsUrl, $newstitle, contrexx_raw2xhtml($newstitle));
+                $newWindow      = empty($objResult->fields['newWindow']) ? "_self" : "_blank";
```
Comment (daeppen) [inline core_modules/News/Controller/News.class.php]: Use single quotes (instead of double quotes) for simple, non-functional strings

File: core_modules/News/Controller/News.class.php [hunk 7]
```diff
@@ -1587,6 +1598,7 @@ private function fetchSubmittedData()
         $data['newsTags'] = !empty($_POST['newsTags'])
             ? contrexx_input2raw($_POST['newsTags'])
             : array();
+        $data['newWindow'] = !empty($_POST['newWindow']) ? $_POST['newWindow'] : 0;
```
Comment (daeppen) [inline core_modules/News/Controller/News.class.php]: Properly sanitize data read from the user as well as data sent to the user. See article [Input / Output Handling](http://wiki.contrexx.com/en/index.php?title=Security#Contrexx_Input_.2F_Output_Handling).

File: core_modules/News/Controller/NewsManager.class.php [hunk 8]
```diff
@@ -3134,7 +3142,8 @@ function createRSS()
                                     'redirect'      => $redirect,
                                     'source'        => $objResult->fields['source'],
                                     'teaser_frames' => explode(';', $objResult->fields['teaser_frames']),
-                                    'categoryIds'    => $this->getNewsRelCategories($objResult->fields['id'])                     
+                                    'categoryIds'   => $this->getNewsRelCategories($objResult->fields['id']),
+                                    'new_window'    => $objResult->fields['newWindow'],
```
Comment (daeppen) [inline core_modules/News/Controller/NewsManager.class.php]: Please explain the purpose of this change/code?

File: core_modules/News/Controller/NewsManager.class.php [hunk 9]
```diff
@@ -1309,6 +1311,7 @@ function add()
                 'TXT_SELECT_ALL'                => $_ARRAYLANG['TXT_SELECT_ALL'],
                 'TXT_DESELECT_ALL'              => $_ARRAYLANG['TXT_DESELECT_ALL'],
                 'TXT_ASSOCIATED_BOXES'          => $_ARRAYLANG['TXT_ASSOCIATED_BOXES'],
+                'TXT_NEW_WINDOW'                => $_ARRAYLANG['TXT_NEW_WINDOW'],
```
Comment (daeppen) [inline core_modules/News/Controller/NewsManager.class.php]: Please follow the scheme of a text-variable as defined in [Interface Text Variable Scheme](http://wiki.contrexx.com/en/index.php?title=Interface_Text#Text-Variable)

File: core_modules/News/View/Template/Backend/module_news_modify.html [hunk 10]
```diff
@@ -472,7 +472,13 @@
                 </tr>
                 <tr class="row1">
                     <td width="12%" nowrap="nowrap"><label for="news_redirect">{TXT_LINK}</label></td>
-                    <td width="88%"><input type="text" name="newsRedirect" id="news_redirect" value="{NEWS_REDIRECT}" style="width:400px;" />&nbsp;<input type="button" value="{TXT_BROWSE}" onClick="newsGetFileBrowser('news_redirect', 'webpages')" />&nbsp;<span class="tooltip-trigger icon-warning"></span><span class="tooltip-message">{TXT_NEWS_TYPE_REDIRECT_HELP}</span></td>
+                    <td width="88%">
+                        <input type="text" name="newsRedirect" id="news_redirect" value="{NEWS_REDIRECT}" style="width:400px;" />&nbsp;
+                        <input type="button" value="{TXT_BROWSE}" onClick="newsGetFileBrowser('news_redirect', 'webpages')" />&nbsp;
+                        <span class="tooltip-trigger icon-warning"></span>
+                        <span class="tooltip-message">{TXT_NEWS_TYPE_REDIRECT_HELP}</span>
+                        <input type="checkbox" name="newWindow" value="1" {NEWS_WINDOW_CHECKED}>{TXT_NEW_WINDOW}
```
Comment (daeppen) [inline core_modules/News/View/Template/Backend/module_news_modify.html]: Please follow naming convention of a HTML attribute as defined in [How to write what](http://wiki.contrexx.com/en/index.php?title=Naming_conventions#How_to_write_what)

Comment (daeppen) [inline core_modules/News/View/Template/Backend/module_news_modify.html]: Add a [tooltip](http://wiki.contrexx.com/en/index.php?title=Tooltip) to help the user understand the purpose/behavior of the function

File: core_modules/News/Controller/NewsManager.class.php [hunk 11]
```diff
@@ -970,6 +970,7 @@ function add()
         $enableRelatedNews      = empty($_POST) || !empty($_POST['enableRelatedNews']) ? 1 : 0;
         $newsTags               = !empty($_POST['newsTags']) ? contrexx_input2raw($_POST['newsTags']) : array();
         $enableTags             = !empty($_POST['enableTags']) ? intval($_POST['enableTags']) : 0;
+        $newWindow             = !empty($_POST['newWindow']) ? intval($_POST['newWindow']) : 0;
```
Comment (daeppen) [inline core_modules/News/Controller/NewsManager.class.php]: Please use proper text indentation

File: core_modules/News/Controller/NewsLibrary.class.php [hunk 12]
```diff
@@ -1217,13 +1217,13 @@ public function parseImageThumbnail($imageSource, $thumbnailSource, $altText, $n
         return array($image, $imageLink, $source);
     }
 
-    protected static function parseLink($href, $title, $innerHtml, $class=null)
+    protected static function parseLink($href, $title, $innerHtml, $target)
```
Comment (daeppen) [inline core_modules/News/Controller/NewsLibrary.class.php]: Define a default value on the new argument to maintain backwards compatibility

**Comment by daeppen** [comment]
@LazuLitee 
Thank you for your contribution!
We're happy to accept your pull-request.
But first, please check my code comments and optimize your code accordingly.
Additionally, please ensure that all calls to **NewsLibrary::parseLink()** are compatible with the new behavior of that method.
Also, to ensure conformity, please also add this new behavior to the following sections:
- NewsHeadlines::getHomeHeadlines()
- NewsLibrary::getMonthlyNewsStats()
- NewsLibrary::getRelatedNewsDetails()

File: core_modules/News/Controller/News.class.php (hunk 1 above)
Comment (LazuLitee) [inline core_modules/News/Controller/News.class.php]: i see... well, none.

File: core_modules/News/Controller/NewsManager.class.php (hunk 8 above)
Comment (LazuLitee) [inline core_modules/News/Controller/NewsManager.class.php]: sorry, there is no explanation. but at least i fixed the indentation.

File: core_modules/News/Controller/News.class.php (hunk 1 above)
Comment (daeppen) [inline core_modules/News/Controller/News.class.php]: - [ ] fixed

File: core_modules/News/Controller/News.class.php (hunk 2 above)
Comment (daeppen) [inline core_modules/News/Controller/News.class.php]: - [ ] fixed

File: core_modules/News/Controller/News.class.php (hunk 3 above)
Comment (daeppen) [inline core_modules/News/Controller/News.class.php]: - [x] fixed

File: core_modules/News/Controller/News.class.php (hunk 4 above)
Comment (daeppen) [inline core_modules/News/Controller/News.class.php]: - [x] fixed

File: core_modules/News/Controller/News.class.php (hunk 5 above)
Comment (daeppen) [inline core_modules/News/Controller/News.class.php]: - [x] fixed

File: core_modules/News/Controller/News.class.php (hunk 6 above)
Comment (daeppen) [inline core_modules/News/Controller/News.class.php]: - [x] fixed

File: core_modules/News/Controller/News.class.php (hunk 7 above)
Comment (daeppen) [inline core_modules/News/Controller/News.class.php]: The sanitation in **empty()** is obsolete:
`!empty(contrexx_input2raw($_POST['redirect_new_window']))`

- [ ] Please remove

File: core_modules/News/lang/de/backend.php [hunk 13]
```diff
@@ -63,7 +63,8 @@
 $_ARRAYLANG['TXT_NEWS_MESSAGE'] = "Newsmeldung";
 $_ARRAYLANG['TXT_EXTERNAL_SOURCE'] = "Externe Quellen";
 $_ARRAYLANG['TXT_LINK'] = "Link";
-$_ARRAYLANG['TXT_NEW_WINDOW'] = "Neues Fenster";
+$_ARRAYLANG['TXT_NEWS_NEW_WINDOW'] = "Neues Fenster";
+$_ARRAYLANG['TXT_NEWS_NEW_WINDOW_HELP'] = "Mithilfe dieser Option können Sie definieren ob der Link der Newsmeldung in einem eigenen Fenster geöffnet werden soll.";
```
Comment (daeppen) [inline core_modules/News/lang/de/backend.php]: - [ ] Please change to "Aktivieren Sie diese Option, damit die Verlinkung in einem neuen Fenster geöffnet wird."

File: core_modules/News/lang/en/backend.php [hunk 14]
```diff
@@ -63,7 +63,8 @@
 $_ARRAYLANG['TXT_NEWS_MESSAGE'] = "News Message";
 $_ARRAYLANG['TXT_EXTERNAL_SOURCE'] = "External sources";
 $_ARRAYLANG['TXT_LINK'] = "Link";
-$_ARRAYLANG['TXT_NEW_WINDOW'] = "New window";
+$_ARRAYLANG['TXT_NEWS_NEW_WINDOW'] = "New window";
+$_ARRAYLANG['TXT_NEWS_NEW_WINDOW_HELP'] = "With this option you can choose if you want to open the news message in a different window.";
```
Comment (daeppen) [inline core_modules/News/lang/en/backend.php]: - [ ] Please change to "Check this option to open the link in a new window."

File: installer/data/contrexx_dump_structure.sql [hunk 15]
```diff
@@ -2520,6 +2520,7 @@ CREATE TABLE `contrexx_module_news` (
   `allow_comments` tinyint(1) NOT NULL DEFAULT '0',
   `enable_related_news` tinyint(1) NOT NULL DEFAULT '0',
   `enable_tags` tinyint(1) NOT NULL DEFAULT '0',
+  `new_window` tinyint(1) NOT NULL DEFAULT '0',
```
Comment (daeppen) [inline installer/data/contrexx_dump_structure.sql]: - [ ] please also change the name of the new attribute to match with the rest of the code: **redirect_new_window**

File: core_modules/News/View/Template/Backend/module_news_modify.html [hunk 16]
```diff
@@ -472,7 +472,14 @@
                 </tr>
                 <tr class="row1">
                     <td width="12%" nowrap="nowrap"><label for="news_redirect">{TXT_LINK}</label></td>
-                    <td width="88%"><input type="text" name="newsRedirect" id="news_redirect" value="{NEWS_REDIRECT}" style="width:400px;" />&nbsp;<input type="button" value="{TXT_BROWSE}" onClick="newsGetFileBrowser('news_redirect', 'webpages')" />&nbsp;<span class="tooltip-trigger icon-warning"></span><span class="tooltip-message">{TXT_NEWS_TYPE_REDIRECT_HELP}</span></td>
+                    <td width="88%">
+                        <input type="text" name="newsRedirect" id="news_redirect" value="{NEWS_REDIRECT}" style="width:400px;" />&nbsp;
+                        <input type="button" value="{TXT_BROWSE}" onClick="newsGetFileBrowser('news_redirect', 'webpages')" />&nbsp;
+                        <span class="tooltip-trigger icon-warning"></span>
+                        <span class="tooltip-message">{TXT_NEWS_TYPE_REDIRECT_HELP}</span>
+                        <input type="checkbox" name="redirect_new_window" value="1" {NEWS_NEW_WINDOW_CHECKED}>{TXT_NEW_WINDOW}
```
Comment (daeppen) [inline core_modules/News/View/Template/Backend/module_news_modify.html]: - [ ] Label of form element must be put into proper HTML-**label** tag

**Comment by daeppen** [comment]
- [ ] Ensure the complete extension works. Currently the new implemented placeholder does not work in most sections
- [ ] The new option must only be applied in case the news article is a redirection

**Comment by LazuLitee** [comment]
I made all corrections, please check.
